crates/cli/src/config.rs

```diff
@@ -109,6 +109,51 @@ pub struct RawConfig {
 pub struct Config {
     proj: RawConfig,
     home: RawConfig,
+
+    /// Lockfile for the project config, if a file exists for the project config.
+    ///
+    /// Created before reading the project config, and closed upon dropping the config.
+    /// Allows us to mutate the config via [`Config::save`] without worrying about other processes.
+    proj_lock: Option<PathBuf>,
+
+    /// Lockfile for the system config.
+    ///
+    /// Created before reading or creating the home config, and closed upon dropping the config.
+    /// Allows us to mutate the config via [`Config::save`] without worrying about other processes.
+    ///
+    /// None only in tests, which are allowed to concurrently mutate configurations as much as they want,
+    /// since they use a hardcoded configuration rather than loading from a file.
+    home_lock: Option<PathBuf>,
```

Comment:
Instead of shuffling around `PathBuf`s, how would you feel about making an actual type `LockFile` or something? It would have its own `drop`, and instead of `create_lockfile_or_panic`, we'd have a function that either panicks or returns a `LockFile`.

Then, it wouldn't just take a `PathBuf` to construct a `Config` - it requires passing a `LockFile`, which represents an actual acquired `LockFile`. (It still doesn't stop you from passing one `LockFile` in place of another.. but it helps prevent passing along a path without remembering to call the lock function)

Reply:
I guess one could also imagine a `LockedConfig` class or some such, rather than bundling the lockfiles into `Config`. It would just tightly couple access to the the `Config` instance, with holding the lockfile for the correct  corresponding path, and when one is dropped the other is dropped. Then we wouldn't be passing along multiple paths/lockfiles with any risk for confusing them with one another (they'd be tightly coupled with their corresponding `Config`s, and share the same lifetime constraints with those configs rather than with one another)

Reply:
It turns out we don't actually need to lock the `proj` config, since we never write to it. So now we only have a single `Lockfile` on the `home` config.